### PR TITLE
New features for Zernikes

### PR DESCRIPTION
**Pull request by jmeyers314**
A quick PR to add some new features to Zernikes.  The one I'm most interested in is the Hessian determinant.  This shows up in the fast analytic donut fitting algorithm here: https://dspace.mit.edu/bitstream/handle/1721.1/78543/836817391-MIT.pdf?sequence=2

To ease the implementation, I added +, -, and * operators (though no / operator).

I also added a Laplacian operator, which shows up in transport-of-intensity type donut algorithms.

### CODE REVIEW

File: galsim/zernike.py [hunk 1]
```diff
@@ -433,6 +438,91 @@ def __init__(self, coef, R_outer=1.0, R_inner=0.0):
         self.R_outer = float(R_outer)
         self.R_inner = float(R_inner)
 
+    def __add__(self, rhs):
```
Comment (rmjarvis) [inline galsim/zernike.py]: Should add doc strings for these arithmetic operations.  We'll want to list them in the Sphinx docs for the class.

File: tests/test_zernike.py [hunk 2]
```diff
@@ -414,6 +418,214 @@ def finite_difference_gradient(Z, x, y):
...
+        z1 * galsim.Gaussian(fwhm=1)
+    with np.testing.assert_raises(ValueError):
+        z1 * galsim.zernike.Zernike([0,1], R_outer=z1.R_outer*2)
+    with np.testing.assert_raises(ValueError):
+        z1 * galsim.zernike.Zernike([0,1], R_outer=z1.R_outer, R_inner=z1.R_inner*2)
+
+    # Commutative with integer coefficients
+    z1 = galsim.zernike.Zernike([0,1,2,3,4,5])
+    z2 = galsim.zernike.Zernike([1,2,3,4,5,6])
+    assert z1*z2 == z2*z1
```
Comment (rmjarvis) [inline tests/test_zernike.py]: I think multiplication by a scalar is only implicitly tested via `__sub__`.  Probably worth adding a real test of that.

File: tests/test_zernike.py [hunk 3]
```diff
@@ -414,6 +418,214 @@ def finite_difference_gradient(Z, x, y):
...
+        )
+        np.testing.assert_allclose(
+            z.hessian.R_inner,
+            R_inner
+        )
+
+    # Do one by hand
+    # Z4 = 2 * sqrt(3) * (x^2 + y^2)
+    # implies hessian = 4*sqrt(3) * 4*sqrt(3) - 0*0 = 16*3 = 48
+    # which is 48 * Z1
```
Comment (rmjarvis) [inline tests/test_zernike.py]: Maybe do one that has a non-trivial d^2z/dxdy term.

File: tests/test_zernike.py (hunk 2 above)
Comment (jmeyers314) [reply tests/test_zernike.py]: This is tested in `test_sum`, but I added it here too to make it easier to find.

**Comment by jmeyers314** [comment]
Thanks Mike, I accepted all your suggestions.  Note, I force-pushed this back up after rebasing on top of the sphinx changes (not for any good reason, just what I did).  I'll plan to merge tomorrow.

**Comment by jmeyers314** [comment]
Actually, one thing I noticed worth mentioning is that when I rebuilt the sphinx docs here to inspect them, I ended up with ~140 new or modified files in the docs/ directory.  I assume we don't want to maintain the complete git history of all the sphinx-generated html/doctree/ttf/etc files, so I didn't commit those (I did commit the docs source files though).  Probably there's a standard way to get these up to github.io without making them part of the source repo?  Orphan branch maybe?

**Comment by rmjarvis** [comment]
I usually rebase when there are conflicts to make the merge back to master straightforward.

File: docs/zernike.rst [hunk 4]
```diff
@@ -4,6 +4,12 @@ Zernike Functions
 .. autoclass:: galsim.zernike.Zernike
     :members:
 
+    .. automethod:: galsim.zernike.Zernike.__add__
+    .. automethod:: galsim.zernike.Zernike.__sub__
+    .. automethod:: galsim.zernike.Zernike.__neg__
+    .. automethod:: galsim.zernike.Zernike.__mul__
+    .. automethod:: galsim.zernike.Zernike.__rmul__
```
Comment (rmjarvis) [inline docs/zernike.rst]: and `__call__`

**Comment by rmjarvis** [comment]
The stuff in _build doesn't need to be committed.  In fact, normally you really shouldn't, since we probably want the Sphinx docs (which have to live on master for GitHub reasons) to be constant at the current release.  So updating in between releases is not desirable.

Right now, we are kind of stuck with having the docs be ahead of the release.  Part of the reason I want to get to 2.2 relatively soon.  (Not the only reason though.)  So for now, it doesn't matter whether you do or don't commit them as part of this merge.

But in general, we'll just update and commit the _build directory when we are cutting a new release.

**Comment by jmeyers314** [comment]
>the Sphinx docs (which have to live on master for GitHub reasons)

We can't host them on [gh-pages](https://www.alkaline-ml.com/2018-12-23-automate-gh-builds/)?

AFAICT, hosting them on master means extra git work cleaning things up whenever one wishes to inspect the sphinx output during development.  Maybe there's a way to set the sphinx output directory to something with a .gitignore during development, and setting it to _build only when cutting a release?

**Comment by rmjarvis** [comment]
We can, but it's more work at release time, IMO, than hosting them on master.  I tried both for TreeCorr, and I found the workflow for putting them on master to be easier than using gh-pages.

I usually just do `git checkout docs/_build` after I'm happy with the docs before committing.  But if you want, we can add the `_build` directory to `.gitignore`.

**Comment by rmjarvis** [comment]
> the Sphinx docs (which have to live on master for GitHub reasons)

BTW, this comment was meant to be that we can't have the docs point to `releases/2.1/docs`, which would have been a convenient solution if it was allowed.  Other than a gh-pages orphan branch, the only branch we are allowed to use is master.

**Comment by jmeyers314** [comment]
I did try `git checkout docs/_build` and but ended up with a bunch of untracked files:

```
[meyers18@Temeria ~/src/GalSim (zernike_enhance|✚142…)]$ git checkout docs/_build
[meyers18@Temeria ~/src/GalSim (zernike_enhance|…)]$ git status
On branch zernike_enhance
Your branch is up to date with 'origin/zernike_enhance'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	docs/_build/html/_downloads/05668572b392713ce989279727b22a2d/
	docs/_build/html/_downloads/0cab3503f6dfcc06ea3328e17970c969/
	docs/_build/html/_downloads/0d827c6a9891b4608e4635e6e6286483/
	docs/_build/html/_downloads/1659fc602878138bcddcaefae3b26282/
	docs/_build/html/_downloads/19a3b0870acf3d06a65cf732e55342b7/
	docs/_build/html/_downloads/1bf7f234031958b2d3047c98b245989e/
	docs/_build/html/_downloads/1df49db2c847fc2e42e31d69fc2782be/
	docs/_build/html/_downloads/1e16dce3d6290b145d5ecad5b692b907/
	docs/_build/html/_downloads/22da0ee745454c26b1efaf814de64fab/
	docs/_build/html/_downloads/2cc1046ae60a3b13fc3674a36a2b9fcb/
	docs/_build/html/_downloads/42fc06f7f3e799ad48a7a8940dd404bc/
	docs/_build/html/_downloads/4554138f5d7b6654a2bc81eec4694576/
	docs/_build/html/_downloads/46830156a556242136f70458eac3030b/
	docs/_build/html/_downloads/4ad3dcd8aa87d39e457824251156b106/
	docs/_build/html/_downloads/4e6d609c331c0ae38f294e4977ef244e/
	docs/_build/html/_downloads/55b14b788e74c8ad6ec6fbb1dd3be214/
	docs/_build/html/_downloads/58f03e35661220c92d8e111d97d5269d/
	docs/_build/html/_downloads/59334d80bb64e9baffc1a834414b4964/
	docs/_build/html/_downloads/5dd4fe2b9882e94a46fb1a0e8cbfffef/
	docs/_build/html/_downloads/62c53e73dc3de0fecdc4648b6266e06e/
	docs/_build/html/_downloads/70947764bebec22b8650c8f7bf1e85c1/
	docs/_build/html/_downloads/7445fe49c10c6e6816b40af2e7735300/
	docs/_build/html/_downloads/7e357062a026cf42bf00add19e035bfb/
	docs/_build/html/_downloads/81417fbce8a289b385cc7fac81f2ac1b/
	docs/_build/html/_downloads/872c7d708ae4d88bce7f20e9e6f6289f/
	docs/_build/html/_downloads/8db0053a71d4970bbf173b33da9351bf/
	docs/_build/html/_downloads/8f47d6c4be2d9deddd1052bf631589e7/
	docs/_build/html/_downloads/9a473e60901c8486d545804d8d0391e8/
	docs/_build/html/_downloads/bfaff1c6e5f0d9037107b75be106ec14/
	docs/_build/html/_downloads/c74cdcd8ab856a081dfce1ac4c92ec53/
	docs/_build/html/_downloads/c8881884aac00ee95006d44e0c75401b/
	docs/_build/html/_downloads/cc01f761e8cc8e3ab4675edaa57f5b45/
	docs/_build/html/_downloads/d9ec0fa814422ed3a48ac932b43d473d/
	docs/_build/html/_downloads/dbd6e9c719463f86cbd0ee357f9edc22/
	docs/_build/html/_downloads/de69a17f2f9096ca7a20438ca0815ef4/
	docs/_build/html/_downloads/e72592776fafb544355792aa73f3e6da/
	docs/_build/html/_downloads/ea3fd391fba5e7cb1a0c2741c27f45a1/
	docs/_build/html/_downloads/f1c3307769ca45baed69a53fde78fe7e/
	docs/_build/html/_downloads/f2ecd62dbba24f9347dea5d0050a68d5/
	docs/_build/html/_downloads/f8eff6ff06ddd71e192285e38b1ee6ab/

nothing added to commit but untracked files present (use "git add" to track)
```

I'll try .gitignoring docs/_build though.

**Comment by rmjarvis** [comment]
I added a `clean_docs` script to (maybe?) make it easier to clean up the docs/_build directory when you don't actually want to commit it.

**Comment by jmeyers314** [comment]
That does seem to work.  I still like the .gitignore though, since it will protect us all from an inevitable `git commit -am"blah blah blah"`.  I'm *usually* more careful than that, but I wouldn't put it past me to make that mistake :) .

**Pull request merged**